File: leetcode/0052.n-quees-2/go/001.backtracking.go
func totalNQueens(n int) int {
    res := 0
    const (
        QUEEN = 'Q'
        EMPTY = '.'
    )

    board := make([][]byte, n)
    for row := range(n) {
        board[row] = make([]byte, n)
        for col := range(n) {
            board[row][col] = EMPTY
        }
    }

    colSet := make(map[int]bool)
    mainDiag := make(map[int]bool) // \ - row-col=const
    secDiag := make(map[int]bool) // / - row+col=const

    var backtrack func(int)
    backtrack = func(row int) {
        if row == n {
            res++
            return
        }

        for col := range(n) {
            if colSet[col] || mainDiag[row-col] || secDiag[row+col] {
                continue
            }

            colSet[col] = true
            mainDiag[row-col] = true
            secDiag[row+col] = true
            board[row][col] = QUEEN

            backtrack(row+1)

            delete(colSet, col)
            delete(mainDiag, row-col)
            delete(secDiag, row+col)
            board[row][col] = EMPTY
        }
    }

    backtrack(0)

    return res
}

File: leetcode/0051.n-queens/ts/001.backtracking.ts
function solveNQueens(n: number): string[][] {
    const EMPTY = '.';
    const QUEEN = 'Q';
    const board = Array(n).fill(EMPTY).map(row => Array(n).fill(EMPTY));
    const res = [];
    const colSet = new Set();
    const mainDiag = new Set(); // \ row - col = const 
    const secDiag = new Set(); // / row + col = const

    const backtrack = (row: number) => {
        if(row === n) {
            res.push(board.map(row => row.join('')));
            return
        }

        for(let col = 0; col < n; col++){
            if(colSet.has(col) || mainDiag.has(row-col) || secDiag.has(row+col)) {
                continue;
            }

            colSet.add(col);
            mainDiag.add(row-col);
            secDiag.add(row+col);
            board[row][col] = QUEEN;

            backtrack(row+1);

            colSet.delete(col);
            mainDiag.delete(row-col);
            secDiag.delete(row+col);
            board[row][col] = EMPTY;
        }
    }

    backtrack(0);

    return res;
};
